src/compas_ags/ags/constraints.py

```diff
@@ -0,0 +1,368 @@
+    -----------
+    form: :class:`FormDiagram`
+
+    References
+    ----------
+    .. [1] Alic, V. and Åkesson, D., 2017. Bi-directional algebraic graphic statics. Computer-Aided Design, 93, pp.26-37.
+
+    """
+
+    def __init__(self, form):
+        self.constraints = []
+        self.form = form
+
+    def add_constraint(self, constraint):
+        self.constraints.append(constraint)
+
+    def compute_constraints(self):
+        jac = np.zeros((0, self.constraints[0].number_of_cols))
+        res = np.zeros((0, 1))
+        for constraint in self.constraints:
+            j, r = constraint.compute_constraint()
+            jac = np.vstack((jac, j))
+            res = np.vstack((res, r))
+        return jac, res
+
+    def update_constraints(self):
+        for constraint in self.constraints:
+            constraint.update_constraint_goal()
+
+    def constraints_from_form(self):
+        """Automate set up of constraint collection based on diagram's attributes"""
+
+        # fix x and y coordinates of the fixed vertices
+        for key in self.form.vertices_where({'is_fixed': True}):
+            self.add_constraint(HorizontalFix(self.form, key))
+            self.add_constraint(VerticalFix(self.form, key))
+
+        self.constrain_dependent_leaf_edges_lengths()
+
+    def get_lines(self):
+        """Get lines to draw in viewer."""
+        lines = []
+        for constraint in self.constraints:
+            constraint_lines = constraint.get_lines()
+            if constraint_lines:
+                lines = lines + constraint_lines
+        return lines
+
+    def constrain_dependent_leaf_edges_lengths(self):
+        leaves = self.form.leaves()
+        dependent_leaf_edges = []
+        for i, (u, v) in enumerate(self.form.edges()):
+            if u in leaves or v in leaves:
+                if not self.form.edge_attribute((u, v), 'is_ind'):
+                    dependent_leaf_edges.append((u, v))
+        for edge in dependent_leaf_edges:
+            self.add_constraint(LengthFix(self.form, edge))
+
+
+class HorizontalFix(AbstractConstraint):
+    """Constraint that keeps x-coordinate of a vertex fixed.
+
+    Parameters
+    -----------
+    form: :class:`FormDiagram`
+        The Form Diagram to constraint
+    vertex: int
+        Idenfifier ofthe vertex to fix.
+
+    """
+
+    def __init__(self, form, vertex):
+        super().__init__(form)
+        self.vertex = vertex
+        self.x = None
+        self.set_initial_position()
+
+    def set_initial_position(self):
+        self.x = self.form.vertex_attribute(self.vertex, 'x')
+
+    def compute_constraint(self):
+        constraint_jac_row = np.zeros((1, self.number_of_cols))
+        idx = self.vertex_index[self.vertex]
+        constraint_jac_row[0, idx] = 1
+        r = self.form.vertex_attribute(self.vertex, 'x') - self.x
+        return constraint_jac_row, r
+
+    def update_constraint_goal(self):
+        self.set_initial_position()
+
+    def get_lines(self):
```

Comment:
Ultimately @brgcode I would suggest we have something like these constraint drawings to rhino->artists... But still we need such a framework to display constraints in pure python....

Reply:
sure but also on the non-Rhino side it would be better to move this functionality to artists and objects of the plotter and/or viewer...